lib/Optimizer/Transforms/DependencyAnalysis.cpp

```diff
@@ -0,0 +1,3148 @@
+    } else if (isTerminator) {
+      newNode = new TerminatorDependencyNode(op, dependencies);
+    } else {
+      newNode = new OpDependencyNode(op, dependencies);
+    }
+
+    // Dnodeid is the next slot of the dnode vector
+    auto id = perOp.size();
+
+    // Add dnodeid attribute
+    OpBuilder builder(op);
+    op->setAttr("dnodeid", builder.getUI32IntegerAttr(id));
+    perOp.push_back(newNode);
+
+    return newNode;
+  }
+
+  /// Returns the dependency node for the defining operation of \p v
+  /// Assumption: defining operation for \p v exists and already has been
+  /// visited.
+  ///
+  /// If \p v a classical value from a different scope, allocates a
+  /// ShadowDependencyNode, and adds it to the frontier of the parent node
+  /// of the use of \p v that is at the same scope as \p v
+  DependencyNode::DependencyEdge visitValue(Value v) {
+    if (auto barg = dyn_cast<BlockArgument>(v))
+      return DependencyNode::DependencyEdge{argMap[barg], 0};
+
+    auto defOp = v.getDefiningOp();
+    assert(defOp &&
+           "Cannot handle value that is neither a BlockArgument nor OpResult");
+
+    auto resultidx = dyn_cast<OpResult>(v).getResultNumber();
+
+    // Since we walk forward through the ast, every value should be defined
+    // before it is used, so we should have already visited defOp,
+    // and thus should have a memoized dnode for defOp, fail if not
+    assert(defOp->hasAttr("dnodeid") && "No dnodeid found for operation");
+
+    auto id = defOp->getAttr("dnodeid").cast<IntegerAttr>().getUInt();
+    auto dnode = perOp[id];
+
+    if (!ifStack.empty() && defOp->getParentOp() != ifStack.back() &&
+        dnode->isQuantumDependent()) {
+      auto opdnode = static_cast<OpDependencyNode *>(dnode);
+      auto shadow_node = new ShadowDependencyNode{opdnode, resultidx};
+
+      auto parent = ifStack.back();
+
+      while (parent->getParentOp() != defOp->getParentOp())
+        parent = parent->getParentOp();
+
+      freeClassicals[parent].insert(shadow_node);
+
+      return DependencyNode::DependencyEdge{shadow_node, resultidx};
+    }
+
+    return DependencyNode::DependencyEdge{dnode, resultidx};
+  }
+
+  /// Cleans up semi-constructed dependency graph when backing out of running
+  /// DependencyAnalysis because of an encountered error
+  void clean() {
+    // TODO: clean up nodes
+    // Adam: can use perOps, have to be careful about nested nodes though
+  }
+
+  unsigned getNumVirtualAllocs() { return vallocs; }
+};
+
+struct DependencyAnalysisPass
+    : public cudaq::opt::impl::DependencyAnalysisBase<DependencyAnalysisPass> {
+  using DependencyAnalysisBase::DependencyAnalysisBase;
+
+  /// DependencyAnalysis constructs a data structure representing the
+  /// quake AST, performs several analyses/optimizations, and then generates
+  /// a new quake AST based on the resulting data structure.
+  ///
+  /// First, the quake AST is walked, constructing a DependencyBlock for the
```

Comment:
Terminology question: Is "quake AST" just another way of referring to the IR? Or are we to take some addition nuance from it?

There isn't actually an abstract syntax tree (AST) in quake. Quake is a structured IR that is an SSA value graph. That graph is cyclic, has explicit and implicit use/def relations, can use symbolic relations, etc.

My wobbly inference is that this terminology is a way of viewing the order of operations in MLIR's basic blocks as not definitive w.r.t. the actual quantum constraints. That property is always true for classical code as well.

Reply:
Yes, it is referring to the IR. Would "quake code" (or perhaps "unordered quake code") be a more accurate phrasing?

Reply:
Assuming I was reading your mind correctly, I guess "quake IR" is what we are walking across, but we're really interested in the dependency information encoded in the IR. So maybe something like "quake dependency graph"? But then that sort of gets confusing with the on-the-side dependency graph being built here, which is also a dependence graph. (Naming stuff is hard.)